include/RAJA/pattern/teams/teams_cuda.hpp

```diff
@@ -416,121 +322,120 @@ struct LoopExecute<cuda_thread_xyz_nested_direct, SEGMENT> {
     const int len2 = segment2.end() - segment2.begin();
     const int len1 = segment1.end() - segment1.begin();
     const int len0 = segment0.end() - segment0.begin();
+
+    for (int bx = internal::get_cuda_dim<DIM0>(blockIdx);
+         bx < len0;
+         bx += internal::get_cuda_dim<DIM0>(gridDim))
     {
-      const int tx = threadIdx.x;
-      const int ty = threadIdx.y;
-      const int tz = threadIdx.z;
-      if (tx < len0 && ty < len1 && tz < len2)
-        body(*(segment0.begin() + tx),
-             *(segment1.begin() + ty),
-             *(segment2.begin() + tz));
+
+      for (int by = internal::get_cuda_dim<DIM1>(blockIdx);
+           by < len1;
+           by += internal::get_cuda_dim<DIM1>(gridDim))
+      {
+
+        for (int bz = internal::get_cuda_dim<DIM2>(blockIdx);
+             bz < len2;
+             bz += internal::get_cuda_dim<DIM2>(gridDim))
+        {
+
+          body(*(segment0.begin() + bx),
+               *(segment1.begin() + by),
+               *(segment2.begin() + bz));
+        }
+      }
     }
   }
 };
 
 
-// perfectly nested cuda loop policies
-struct cuda_block_xy_nested_loop;
-struct cuda_block_xyz_nested_loop;
 
-struct cuda_thread_xy_nested_loop;
-struct cuda_thread_xyz_nested_loop;
 
-template <typename SEGMENT>
-struct LoopExecute<cuda_block_xy_nested_loop, SEGMENT> {
 
-  template <typename BODY>
+template <typename SEGMENT, int DIM>
+struct TileExecute<cuda_thread_xyz_loop<DIM>, SEGMENT> {
+
+  template <typename TILE_T, typename BODY>
   static RAJA_INLINE RAJA_DEVICE void exec(
       LaunchContext const RAJA_UNUSED_ARG(&ctx),
-      SEGMENT const &segment0,
-      SEGMENT const &segment1,
+      TILE_T tile_size,
+      SEGMENT const &segment,
       BODY const &body)
   {
-    const int len1 = segment1.end() - segment1.begin();
-    const int len0 = segment0.end() - segment0.begin();
+
+    const int len = segment.end() - segment.begin();
+
+    for (int tx = internal::get_cuda_dim<DIM>(threadIdx) * tile_size;
```

Comment:
Nice! Something we have been wanting to do add is the ability to block-out unused threads in the thread loops. For example when we have 3D thread block but we only want to run with threads in the x dimension. Is that something we can add in this PR? Or should we add in a following PR? The thought would just to add in in the nested loop policies. 

The idea here would just add additional if(threadIdx.y < 1) , if(threadIdx.z < 1) for the threaded loops.

Reply:
@artv3  just follow up here:  Do you mean "loop over x, but not over y or z"... like an exclusive x loop?
Like cuda_exclusive_thread_x_loop?
That would be relatively easy to implement.